### PR TITLE
cmake: Add function for adding snippets to the linker file

### PR DESCRIPTION
**Pull request by oyvindronningstad**
Allow using Cmake to place snippets into the linker file.

zephyr_linker_sources() allows injecting linker file fragments into different locations in the linker file:

- Inside the noinit, rwdata and rodata output sections
- Inside the RAM group, for custom RAM output sections.
- Near the end of the linker script, for other custom output sections.

The rationale for this change:

- Keep the linker definitions of a module together with everything else to do with that module.
- Make it less invasive to add a section, since you're less likely to interfere with what is already there.
- Right now it's hard/impossible to know if you are messing with the surrounding sections. It needs to be easier to see what sections/symbols must be placed in a specific position globally.
- Make the linker script cleaner and more readable.
- Support injecting multiple out-of-tree LD script snippets.

This PR deprecates the custom-\*.ld and soc-\*.ld mechanisms which are similar, but more cumbersome to use.

### CODE REVIEW

**Comment by oyvindronningstad** [comment]
> Found the following issues, please fix and resubmit:
> 
> ## Codeowners issues
> ```
> New files added that are not covered in CODEOWNERS:
> 
> arch/common/isr_tables.ld
> arch/common/nocache.ld
> arch/common/ramfunc.ld
> subsys/testsuite/coverage/coverage_ram.ld
> subsys/testsuite/coverage/coverage_rom.ld
> 
> Please add one or more entries in the CODEOWNERS file to cover those files
> ```
> ## License issues
> In most cases you do not need to do anything here, especially if the files
> reported below are going into ext/ and if license was approved for inclusion
> into ext/ already. Fix any missing license/copyright issues. The license
> exception if a JFYI for the maintainers and can be overriden when merging the
> pull request.
> 
> * arch/common/isr_tables.ld missing license.
> * arch/common/isr_tables.ld missing copyright.
> * subsys/testsuite/coverage/coverage_ram.ld missing license.
> * subsys/testsuite/coverage/coverage_ram.ld missing copyright.

Not quite sure what to do with this. The linker code was copied verbatim from linker.ld.

File: arch/arm/core/cortex_m/vt_pointer_section.ld [hunk 1]
```diff
@@ -0,0 +1,6 @@
+/* Reserved 4 bytes to save vector table base address */
+  SECTION_PROLOGUE(.vt_pointer,(NOLOAD),)
+  {
+          *(.vt_pointer_section)
+          *(\".vt_pointer_section.*\")
```
Comment (SebastianBoe) [inline arch/arm/core/cortex_m/vt_pointer_section.ld]: Should omit the backslashes, right?

File: Kconfig.zephyr [hunk 2]
```diff
@@ -149,24 +149,6 @@ config CUSTOM_LINKER_SCRIPT
...
-	help
-	  Include a customized linker script fragment for inserting additional
-	  data and linker directives into the data section.
-
-config CUSTOM_SECTIONS_LD
-	bool "Include custom-sections.ld"
-	help
-	  Include a customized linker script fragment for inserting additional
-	  arbitrary sections.
-
```
Comment (SebastianBoe) [inline Kconfig.zephyr]: Before dropping support for CUSTOM_RODATA_LD we want to first mark it as deprecated to give users a reasonable window of time to migrate.

The same applies to SOC.

But we should still migrate the in-tree users.

**Review by SebastianBoe — APPROVED** [review]
IMHO the rationale is valid and we should adopt this feature.

It has been a long-standing issue for general out-of-tree support that we constantly need to patch the in-tree linker script.

File: Kconfig.zephyr (hunk 2 above)
Comment (oyvindronningstad) [reply Kconfig.zephyr]: Is there a way to deprecate beyond mentioning it in the prompt/help text?

Comment (SebastianBoe) [reply Kconfig.zephyr]: Unfortunately not.

Comment (oyvindronningstad) [reply Kconfig.zephyr]: Take a look now

Comment (SebastianBoe) [reply Kconfig.zephyr]: LGTM :)

File: arch/arm/core/cortex_m/vt_pointer_section.ld (hunk 1 above)
Comment (oyvindronningstad) [reply arch/arm/core/cortex_m/vt_pointer_section.ld]: Yep, fixed, thanks

File: arch/common/CMakeLists.txt [hunk 3]
```diff
@@ -12,3 +12,9 @@ zephyr_sources_ifdef(
   CONFIG_EXECUTION_BENCHMARKING
   timing_info_bench.c
   )
+
+
+zephyr_linker_sources_ifdef(CONFIG_ARCH_HAS_RAMFUNC_SUPPORT
```
Comment (erwango) [inline arch/common/CMakeLists.txt]: Trying to understand consequences and possibilities this offers.
So, suppose a soc requires custom ramfunc support, I'd just have to:
-not to select ARCH_HAS_RAMFUNC_SUPPORT
-add the equivalent in of this code in soc/x/y/CMakeLists.txt
-provide the ad hoc ramfunc.ld where it makes sense
Is that correct understanding?

Comment (oyvindronningstad) [reply arch/common/CMakeLists.txt]: Yes, that's the gist of it. An alternate solution would be to ifdef in the existing ramfunc.ld.

Comment (erwango) [reply arch/common/CMakeLists.txt]: Well, this is more than welcome, then! Thanks.
Let me look have a look to other parts and see if it solves few other pending points.

**Comment by erwango** [comment]
^^@vanti

**Comment by vanti** [comment]
In `include/arch/arm/cortex_m/scripts/linker.ld`, would it be possible to add a snippet insertion point to the beginning of the code section? It seems multiple platforms have felt the need to add something custom to that point:
```

    SECTION_PROLOGUE(_TEXT_SECTION_NAME,,)
        {
#ifdef CONFIG_CC3220SF_DEBUG
        /* Add CC3220SF flash header to disable flash verification */
        . = 0x0;
        KEEP(*(.dbghdr))
        KEEP(*(".dbghdr.*"))
#endif

#ifdef CONFIG_NXP_IMX_RT_BOOT_HEADER
        KEEP(*(.boot_hdr.conf))
        . = CONFIG_IMAGE_VECTOR_TABLE_OFFSET;
        KEEP(*(.boot_hdr.ivt))
        KEEP(*(.boot_hdr.data))
#ifdef CONFIG_DEVICE_CONFIGURATION_DATA
        KEEP(*(.boot_hdr.dcd_data))
#endif
#endif
```
 
In general, it might be interesting to study the linker.ld files and see if there are other points where platform-specific insertions are performed, and consider adding snippet insertion points to those places to help simplify the linker files.

**Comment by ruuddw** [comment]
Looks like an improvement. The better abstractions might help in supporting toolchains other than GNU (with different linker command syntax), like e.g. ARC MetaWare - though more will be needed for that.

**Comment by oyvindronningstad** [comment]
> In `include/arch/arm/cortex_m/scripts/linker.ld`, would it be possible to add a snippet insertion point to the beginning of the code section? It seems multiple platforms have felt the need to add something custom to that point:

I was hesitant about adding this since the order is essential when moving the location counter, but I agree it would be a good addition. I also played with the thought of having an address-aware mechanism for placing things at the start, which would sort the snippets correctly. If that is something more people want, I could try to come up with something.

> In general, it might be interesting to study the linker.ld files and see if there are other points where platform-specific insertions are performed, and consider adding snippet insertion points to those places to help simplify the linker files.

Definitely, if there are other places that make sense for insertion, we should do that.

**Review by erwango — APPROVED** [review]

**Comment by oyvindronningstad** [comment]
@carlescufi What should I do about the license and codeowners issues? Also, should anyone else review this?

**Comment by oyvindronningstad** [comment]
@andrewboie @nashif FYI you have been added as code owners of some of the new files here.

**Review by andyross — COMMENTED** [review]
Cranky drive-by complaint having not read the full patch.

The linker scripts are absolutely one of the messiest areas of our build.  The need to generate/preprocess them is clear, and would absolutely save a ton of typing and maintenance time.  I buy the value of this patch.

But given the need for generation and preprocessing, why on earth are you trying to do this in cmake and not... y'know, python?  I mean, for the current case of simple template, it's fine.  But how do you extend this to do things like:

* Generate static arrays at runtime for the zillion use cases from a single in-source declaration?
* Cross things like those arrays with the set of static app spaces and generate ones for each?
* Add per-arch hooks for things like alignment (page/cacheline/word)

Broadly, its seems like "adding snippets to the linker file" is maybe a 30% solution to a much bigger problem we want to solve, and it's not clear to me how adding this facility helps us get there.

**Comment by wentongwu** [comment]
I like andyross's idea.

**Comment by oyvindronningstad** [comment]
> Cranky drive-by complaint having not read the full patch.

Thanks for the warning ;). I do appreciate the input!

> The linker scripts are absolutely one of the messiest areas of our build. The need to generate/preprocess them is clear, and would absolutely save a ton of typing and maintenance time. I buy the value of this patch.
> 
> But given the need for generation and preprocessing, why on earth are you trying to do this in cmake and not... y'know, python? I mean, for the current case of simple template, it's fine. But how do you extend this to do things like:
> 
> * Generate static arrays at runtime for the zillion use cases from a single in-source declaration?
> * Cross things like those arrays with the set of static app spaces and generate ones for each?
> * Add per-arch hooks for things like alignment (page/cacheline/word)

I don't fully understand your examples, but I would like to hear your opinion after you read the whole patch. Note that there is no generation of linker code, only copying from an existing file. Note also that the Cmake is only used to stitch it together (decide source and destination files). I think a solution with Python would need about as much Cmake code to glue it into the system.

> Broadly, its seems like "adding snippets to the linker file" is maybe a 30% solution to a much bigger problem we want to solve, and it's not clear to me how adding this facility helps us get there.

What we could do is move all this stuff to build time, so proper dependencies can be set up. This means the source files can be generated by a script, which would help with building more sophisticated solutions.

**Comment by oyvindronningstad** [comment]
> @andrewboie @nashif FYI you have been added as code owners of some of the new files here.

I removed my changes to the CODEOWNERS since the file was updated in master and now covers the new files.

**Comment by SebastianBoe** [comment]
> Add per-arch hooks for things like alignment (page/cacheline/word)

This can be done through Kconfig, or, if necessary, through #including <alignment.ld> and modifying the LD include paths appropriately. Each alignment.ld could define it's own macro's for how to do alignment.

> Generate static arrays at runtime for the zillion use cases from a single in-source declaration?
> Cross things like those arrays with the set of static app spaces and generate ones for each?

This is not related in any way to what we are doing here. Sounds like you want some python code and a binutils toolchain.

**Comment by andyross** [comment]
To be clear: my review isn't really a -1, it's a note that long term, we need a more or less complete rewrite of the linkage layer of the build.  And IMHO that's going to rely in large part on programmatically-generated linker scripts.  Given that... I guess I don't see much value to this kind of half effort that preprocesses linker scripts with what amounts to a simple macro language.  It's very likely that any work using this is going to need to be ripped out and redone anyway.

**Comment by oyvindronningstad** [comment]
> To be clear: my review isn't really a -1, it's a note that long term, we need a more or less complete rewrite of the linkage layer of the build. And IMHO that's going to rely in large part on programmatically-generated linker scripts. Given that... I guess I don't see much value to this kind of half effort that preprocesses linker scripts with what amounts to a simple macro language. It's very likely that any work using this is going to need to be ripped out and redone anyway.

I get that. We need something more. My view is that this is, as you said, 30% of the features, for ~5% of the work so I think it's worth bringing it in until something better comes along.

**Comment by vonhust** [comment]
It's an improvement.  But i agreeed with @andyross @ruuddw , if possible, we need a solution with more flexibity and abstraction.  Using python with yaml/json to design a linkscript generator (also can add some check function) is one possible approach. Esepically considering the support of non-gnu toolchain which have different linker script mechnisim,  and applicaitons like TEE.

Anyway, i will vote for this finally as it's indeed an improvement and avoid the hack of generic link script template directly.

**Review by carlescufi — APPROVED** [review]
I agree that this only addresses part of the issues we have today with the linker script, but those are issues that can be addressed now without any real impact to existing applications and are beneficial for out of tree users in the right here right now.

**Comment by SebastianBoe** [comment]
@nashif : All review comments have been addressed, this is ready to be merged / reviewed.

**Review by vonhust — APPROVED** [review]

**Pull request merged**